Rewrite tag based cache invalidation chapter (#638)
contao/docs: docs/dev/framework/caching.md

#### Cache tag invalidation within the Contao back end

When working with DCAs in the Contao back end you don't have to register callbacks at various places to make sure certain
tags are being invalidated. This is because Contao invalidates a certain set of tags whenever a back end entry is created, 
updated or deleted. 

The tags are as follows:

* `contao.db.<table-name>.<id>` (The record itself)
* `contao.db.<table-name>` (Only if the DCA has no parent table defined)

If the DCA has a **parent table**, Contao recursively iterates upwards the table hierarchy and invalidates the following 
tags as well:

> * `contao.db.<parent-table-name>` (Only for the top most parent table)

Reviewer: ```suggestion
* `contao.db.<parent-table-name>` (Only for the topmost parent table)
```

* `contao.db.<parent-table-name>.<pid>`

If the DCA has one or many **child tables**, Contao recursively iterates downwards the table hierachy and invalidates the 
following tags as well:

* `contao.db.<child-table-name>.<cid>`

##### Example: Edit a news article

> Imagine you have edited a news article with ID 42. Contao will now automatically send an invalidation request to the 

Reviewer: ```suggestion

Imagine you have edited a news article with ID 42. Contao will now automatically send an invalidation request to the 
```

reverse proxy to invalidate all responses associated with the following tags: 

> * `contao.db.tl_news_archive` (The parent table)

Reviewer: ```suggestion
* `contao.db.tl_news_archive` (The topmost parent table)
```

* `contao.db.tl_news_archive.1` (The parent record)
* `contao.db.tl_news.42` (The record itself)
* `contao.db.tl_content.420` (The first child record)
* `contao.db.tl_content.421` (The second child record)

> Only the top most parent table tag will be invalidated, i.e. `contao.db.tl_news_archive`, but not `contao.db.tl_news` 

Reviewer: ```suggestion
Only the topmost parent table tag will be invalidated, i.e. `contao.db.tl_news_archive`, but not `contao.db.tl_news` 
```

tags are being invalidated. This is because Contao invalidates a certain set of tags whenever a back end entry is created,
updated or deleted. The tags are as follows:

* `contao.db.<table-name>.<id>`

> * `contao.db.<table-name>` OR `contao.db.<parent-table-name>` (depending on whether a parent table is defined or not)

Reviewer: That's not correct, imho. Correct would be

* `contao.db.<table-name>` in case the current table has no `ptable` and you are thus on top level.
* `contao.db.<parent-table-name>.<pid>` in case the current table has a `ptable` defined.

Reviewer: Basically it goes up and down the tree recursively and adds a general tag for the top level.

So in case of page -> article -> content element if you edit article ID 42 you will end up having the following tags invalidated:

* `contao.db.tl_page` // General tag because there's no ptable for `tl_page`
* `contao.db.tl_page.<page_id>` // page ID = `tl_article.pid`
* `contao.db.tl_article.42`
* `contao.db.tl_content.<ce_id_1>` // All tl_content with ptable `tl_article` and `pid=42`
* `contao.db.tl_content.<ce_id_2>` // All tl_content with ptable `tl_article` and `pid=42`

Author: Thanks for your comment, will check on that 👍

Author: I addressed your comment in my last commit. The `tl_news` example was exactly like your article example, so that should be fine. I broke up the tags list in the record/table itself, parent and child tables and addressed the recursion, to make it more clear.

* `contao.db.<parent-table-name>.<pid>` (only if there is a parent record)

> * `contao.db.<child-table-name>.<cid>` (only if there is a child record)

Reviewer 2: ```suggestion
* `contao.db.<child-table-name>.<cid>` (only if there are child records)
```

Author: Updated accordingly

Imagine you  edited a news article with ID 42. Contao will now automatically send an invalidation request to the 
reverse proxy to invalidate all responses associated with the following tags: 

* `contao.db.tl_news.42`
* `contao.db.tl_news_archive`
* `contao.db.tl_news_archive.1`

> * `contao.db.tl_content.420`

Reviewer 2: ```suggestion
* `contao.db.tl_content.420`
* `contao.db.tl_content.421`
```
I would add another content element here to make it clear that there can be multiple of them.

Author: Updated accordingly
